crypto/evp/pmeth_gn.c

```diff
@@ -35,18 +35,22 @@ static int gen_init(EVP_PKEY_CTX *ctx, int operation)
     if (ctx->keymgmt == NULL || ctx->keymgmt->gen_init == NULL)
         goto legacy;
 
-    switch (operation) {
-    case EVP_PKEY_OP_PARAMGEN:
-        ctx->op.keymgmt.genctx =
-            evp_keymgmt_gen_init(ctx->keymgmt,
-                                 OSSL_KEYMGMT_SELECT_ALL_PARAMETERS, NULL);
-        break;
-    case EVP_PKEY_OP_KEYGEN:
-        ctx->op.keymgmt.genctx =
-            evp_keymgmt_gen_init(ctx->keymgmt, OSSL_KEYMGMT_SELECT_KEYPAIR,
-                                 NULL);
-        break;
-    }
+    if (selection == 0)
+        switch (operation) {
+        case EVP_PKEY_OP_PARAMGEN:
+            ctx->op.keymgmt.genctx =
+                evp_keymgmt_gen_init(ctx->keymgmt,
+                                     OSSL_KEYMGMT_SELECT_ALL_PARAMETERS, NULL);
+            break;
+        case EVP_PKEY_OP_KEYGEN:
```

Comment:
It seems simpler to keep the existing block of code here, and simply replace `OSSL_KEYMGMT_SELECT_KEYPAIR` with `selection`.